### ISSUE TITLE
[proposal] Add E2E tests for koordlet

### ISSUE THREAD

**Issue by saintube** [koordinator-sh/koordinator#1704]
**What is your proposal**:

Currently, the end-to-end test for the koordlet is missing. Though the basic metrics reporting and batch overcommitment are tested in the slo-controller's E2E, we expect to check the QoS strategies provided by the koordlet like BECPUSuppress, BEMemoryEvict, BECPUEvict, CPUBurst, CPU QoS/Memory QoS, and so on.

**Why is this needed**:

**Is there a suggested solution, if so, please add it**:

After #1573 and #1688, the E2E tests are running on the CI pipeline of GitHub Actions. We can enable to run new E2E tests by adjusting the commands in `.github/workflows/e2e-k8s-xxx.yaml`.

**Comment by aclfe** [comment]
我来帮忙把这些 TODO 的 tests 完成
多谢

**Comment by ZiMengSheng** [comment]
## Progress ledger (maintainer triage, Sep 2026)

Recording state here so contributors don't duplicate work — this proposal has attracted **8 PRs** and they all create the same missing scaffolding.

### Current state on main
`test/e2e/koordlet/` **does not exist yet**. Existing E2E suites are only `test/e2e/slocontroller`, `scheduling`, and `quota`. Every koordlet E2E PR therefore has to bring its own framework/suite registration, which means **they conflict with each other by construction**.

### Ordering dependency (important for contributors and reviewers)
**#3006 should land first** — it is the only open PR that includes `test/e2e/koordlet/framework.go` (the shared harness) alongside its BECPUSuppress + BECPUEvict cases. The remaining PRs should then rebase on top of it and only add their own test files:

| PR | Coverage | Status |
|---|---|---|
| **#3006** | BECPUSuppress + BECPUEvict **+ framework.go** | land first (base) |
| #3012 | CPU QoS (cpu.shares / cpuset isolation) | rebase onto #3006 |
| #3014 | BlkIOQoS | rebase onto #3006 |
| #3016 | resctrl | rebase onto #3006 |

### Closed for inactivity — code still reusable
These were closed in a stale sweep (authors inactive since May 2026), but the work is **not obsolete** and none of it is covered on main. Anyone picking up a gap below is encouraged to start from these diffs:

- **BEMemoryEvict** → #3002
- **BECPUSuppress** → #3007 (superseded by #3006, useful as a scenario reference)
- **psi_collector** → #3008
- **HugePageReport** → #3021

### Remaining gaps (no PR, open for contribution)
Memory QoS, CPUBurst, GroupIdentity, and the runtime-hook paths still have no E2E coverage or open PR.

Keeping this issue open with `help wanted` / `good first issue` — it is a genuinely valuable long-running effort, not a dormant proposal.

**Comment by ZiMengSheng** [comment]
### Addendum to the ordering table: a wiring requirement that applies to all koordlet e2e PRs

While pre-reviewing #3006 (the base PR) I found a defect that would otherwise have been inherited by every PR in this effort, so recording it here rather than only on that PR.

**None of the koordlet e2e tests written so far would actually execute.** Two hookups are missing, and neither is caught by CI — the specs simply never register, so CI goes green while the tests are inert:

1. **`test/e2e/e2e_test.go` needs a blank import.** Suites are linked in explicitly (see lines 45-47 for `quota` / `scheduling` / `slocontroller`). Without `_ "github.com/koordinator-sh/koordinator/test/e2e/koordlet"`, the package is never linked and `ginkgo.Describe` never runs.

2. **The workflow focus filter needs `koordlet`.** `.github/workflows/e2e-k8s-1.28.yaml:113` runs with `--focus='slo-controller|\] Reservation|\] Preemption'`. Anything tagged `[koordlet]` is filtered out even after it registers. Same for the other `e2e-k8s-*.yaml` files.

Since #3006 is the base, it should carry both changes; #3012 / #3014 / #3016 then inherit them for free after rebasing. **Contributors picking this up: adding a test file alone is not sufficient — verify your specs appear in the CI ginkgo output, not just that CI is green.**

Good news on the environment side: `hack/deploy_kind.sh` does deploy koordlet as a DaemonSet and the workflow blocks until it is ready, so cgroup-level assertions against a live koordlet are workable in kind. The prerequisites are in place; only the wiring is missing.

I've asked @saintube to weigh in on whether the shared base should absorb the config-patch and cgroup-poll helpers now, before three PRs rebase onto it — changing the base shape afterwards is the expensive path.
